internal/jobservice/repository/job_service_in_memory.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	return &jobResponse.jobResponse, nil
+}
+
+// Update in memory JobStatus Map with jobId and our JobTable
+func (inMem *InMemoryJobServiceRepository) UpdateJobServiceDb(jobId string, jobTable *JobTable) error {
+	log.Infof("Updating JobId %s with State %s", jobId, jobTable.jobResponse.State)
+	inMem.jobStatus.jobLock.Lock()
+	defer inMem.jobStatus.jobLock.Unlock()
+	jobResponse, ok := inMem.jobStatus.jobMap[jobId]
+	if ok && jobResponse.jobResponse.State == jobTable.jobResponse.State {
+	} else {
+
+		inMem.jobStatus.jobMap[jobId] = jobTable
+	}
+	return nil
+}
+
+// Health Check for this service
+// True always in the in memory case
+func (inMem *InMemoryJobServiceRepository) HealthCheck() bool {
+	return true
+}
+
+// Check if JobSet is in our JobSetMap
+func (inMem *InMemoryJobServiceRepository) IsJobSetSubscribed(jobSetId string) bool {
+	inMem.jobStatus.subscribeLock.Lock()
+	defer inMem.jobStatus.subscribeLock.Unlock()
+	_, ok := inMem.jobStatus.subscribeMap[jobSetId]
+	return ok
+}
+
+// Mark our JobSet as being subscribed
+// SubscribeTable contains JobSet and time when it was created.
+func (inMem *InMemoryJobServiceRepository) SubscribeJobSet(jobSetId string) {
+	inMem.jobStatus.subscribeLock.Lock()
+	defer inMem.jobStatus.subscribeLock.Unlock()
+	_, ok := inMem.jobStatus.subscribeMap[jobSetId]
+	if ok {
+		return
+	} else {
+		inMem.jobStatus.subscribeMap[jobSetId] = NewSubscribeTable(jobSetId)
+	}
+
+}
+
+// UnSubscribe to JobSet and delete all the jobs in the in memory map
+func (inMem *InMemoryJobServiceRepository) UnSubscribeJobSet(jobSetId string) {
+	inMem.jobStatus.subscribeLock.RLock()
+	defer inMem.jobStatus.subscribeLock.RUnlock()
+	_, ok := inMem.jobStatus.subscribeMap[jobSetId]
+	if !ok {
+		log.Infof("JobSetId %s already unsubscribed", jobSetId)
+		return
+	}
+	delete(inMem.jobStatus.subscribeMap, jobSetId)
+	log.Infof("JobSetId %s unsubscribed", jobSetId)
+	inMem.DeleteJobsInJobSet(jobSetId)
+}
+
+// Checks JobSet table to make determine if we should unsubscribe from JobSet
+// configTimeWithoutUpdates is a configurable value that is read from the config
+// We allow unsubscribing if the jobset hasn't been updated in configTime
+func (inMem *InMemoryJobServiceRepository) CheckToUnSubscribe(jobSetId string, configTime int64) bool {
+	if !inMem.IsJobSetSubscribed(jobSetId) {
+		return false
+	}
+	currentTime := time.Now().Unix()
+	for _, val := range inMem.jobStatus.subscribeMap {
+		if val.subscribedJobSet == jobSetId {
+			if (currentTime - val.lastRequestTimeStamp) > configTime {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Update JobSet Map with time that a Job in that JobSet was requested
+func (inMem *InMemoryJobServiceRepository) UpdateJobSetTime(jobSetId string) error {
+	inMem.jobStatus.subscribeLock.Lock()
+	defer inMem.jobStatus.subscribeLock.Unlock()
+
+	_, ok := inMem.jobStatus.subscribeMap[jobSetId]
+	if ok {
+		inMem.jobStatus.subscribeMap[jobSetId] = NewSubscribeTable(jobSetId)
```

Comment:
This is a bit confusing- the function is called `UpdateJobSetTime` but it looks like we replace *all* the existing data associated with the jobset in the subscriptionMap.

Reply:
Actually I'm really confused here- the `subscribeMap` is keyed by `jobSetId` but the object it contains has a `jobId`?